Feature/8950 settings

danielebogo opened the pull request:
Third part of #8950

This PR adds the possibility to manage the site notifications subscription from the settings section.

![animation_3](https://user-images.githubusercontent.com/912252/39757273-3dd81754-52c4-11e8-8f4e-50073d2fc0f4.gif)   ![animation_4](https://user-images.githubusercontent.com/912252/39757274-3df493e8-52c4-11e8-9284-56111c31bc67.gif)

This PR includes:

- The `NotificationSettingsViewController` has been modified to include the _followed sites list_.
- A new view controller called `NotificationSiteSubscriptionViewController` has been created to handle a followed site notifications subscriptions.
- A new table view cell called `CheckmarkTableViewCell` has been created to simply handle the checkmark accessory type. The cell is a subclass of `WPTableViewCell`

**Note:**
I followed the same approach used with `NotificationSettingDetailsViewController` to build the new view controller, using a `UITableViewController` subclass without separate responsibilities in different files about datasource and delegate.
If there's a specific approach to handle tables and collections pleas let me know and i'll fix it.  


**To test:**

- Expand/Close the followed sites list section if the user contains more than 3 sites.
- Update followed site subscriptions using new settings view controller

aerych commented:
Hiya @danielebogo 👋 
I've started peeking at the PR but I'm getting a crash when I tap "View All" to expand either of the list of blogs. ( I have a bunch of sites and I follow a bunch more. )  I haven't checked the code but from the stack trace it looks like there is some trouble updating the sections. 
Details are as follows:
### Stack trace
```
2018-05-08 13:06:41.184966-0500 WordPress[10227:127508] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid update: invalid number of rows in section 0.  The number of rows contained in an existing section after the update (11) must be equal to the number of rows contained in that section before the update (4), plus or minus the number of rows inserted or deleted from that section (0 inserted, 0 deleted) and plus or minus the number of rows moved into or out of that section (0 moved in, 0 moved out).'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001195841e6 __exceptionPreprocess + 294
	1   libobjc.A.dylib                     0x0000000118ae7031 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000119589472 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x0000000115c0a652 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 193
	4   UIKit                               0x0000000110f32dfb -[UITableView _endCellAnimationsWithContext:] + 19031
	5   UIKit                               0x0000000110f4cf6a -[UITableView _updateSections:withUpdateAction:rowAnimation:headerFooterOnly:usingPresentationValues:] + 970
	6   UIKit                               0x0000000110f4d3de -[UITableView reloadSections:withRowAnimation:] + 140
	7   WordPress                           0x000000010e03cfe9 _T09WordPress34NotificationSettingsViewControllerC17toggleDisplayMore33_A87F670D4E857A88181ADCB50058E1ABLLy10Foundation9IndexPathV2at_tF + 361
	8   WordPress                           0x000000010e039e6c _T09WordPress34NotificationSettingsViewControllerC05tableE0ySo07UITableE0C_10Foundation9IndexPathV014didSelectRowAtjK0tF + 140
	9   WordPress                           0x000000010e03a04c _T09WordPress34NotificationSettingsViewControllerC05tableE0ySo07UITableE0C_10Foundation9IndexPathV014didSelectRowAtjK0tFTo + 92
	10  UIKit                               0x0000000110f4fee9 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 1813
	11  UIKit                               0x0000000110f50104 -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 344
	12  UIKit                               0x0000000110e1c513 _runAfterCACommitDeferredBlocks + 318
	13  UIKit                               0x0000000110e0b77e _cleanUpAfterCAFlushAndRunDeferredBlocks + 388
	14  UIKit                               0x0000000110e39f05 _afterCACommitHandler + 137
	15  CoreFoundation                      0x0000000119526607 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	16  CoreFoundation                      0x000000011952655e __CFRunLoopDoObservers + 430
	17  CoreFoundation                      0x000000011950ab81 __CFRunLoopRun + 1537
	18  CoreFoundation                      0x000000011950a30b CFRunLoopRunSpecific + 635
	19  GraphicsServices                    0x000000011d166a73 GSEventRunModal + 62
	20  UIKit                               0x0000000110e110b7 UIApplicationMain + 159
	21  WordPress                           0x000000010d59fbd0 main + 144
	22  libdyld.dylib                       0x000000011a982955 start + 1
	23  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

### How the UI looks:
![simulator screen shot - iphone 8 plus - 2018-05-08 at 13 04 55](https://user-images.githubusercontent.com/1435271/39774876-cf99a098-52c1-11e8-94b0-a606182f49c8.png)

Let me know if I can provide any other details to help reproduce the issue.

danielebogo commented:
Hi @aerych thanks for the stack trace. I fix the problem with the section reload and now should be ok. I also updated the logic to display the different sections.

aerych commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationSettingsViewController.swift:
```diff
@@ -364,11 +474,15 @@ open class NotificationSettingsViewController: UIViewController {
-    fileprivate var groupedSettings: [[NotificationSettings]]?
+
+    fileprivate var groupedSettings: [Section: [NotificationSettings]]?
     fileprivate var displayMoreWasAccepted          = false
```
You might consider refactoring how `displayMoreWasAccepted` works, or maybe use a different flag for the followed sites list rather than sharing a single flag between both lists.  When there is multiple blogs in both lists it looks like having a single flag is going to end up expanding both lists (assuming there is no exception).

aerych commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationSettingsViewController.swift:
```diff
@@ -132,19 +150,21 @@ open class NotificationSettingsViewController: UIViewController {
         case .blog where requiresBlogsPagination:
             return displayMoreWasAccepted ? rowCountForBlogSection + 1 : loadMoreRowCount
+        case .followedSites:
+            return displayMoreWasAccepted ? rowCountForFollowedSite + 1 : rowCountForFollowedSite
```
Should 
```return displayMoreWasAccepted ? rowCountForFollowedSite + 1 : rowCountForFollowedSite```
actually be
```return displayMoreWasAccepted ? rowCountForFollowedSite + 1 : loadMoreRowCount```
?

danielebogo replied:
Yes i fixed it in my last commit.

danielebogo commented on WordPress/Classes/Services/ReaderPostService.m:
```diff
@@ -21,6 +21,7 @@
 NSUInteger const ReaderPostServiceMaxSearchPosts = 200;
 NSUInteger const ReaderPostServiceMaxPosts = 300;
 NSString * const ReaderPostServiceErrorDomain = @"ReaderPostServiceErrorDomain";
+NSString * const ReaderPostServiceToggleSiteFollowingState = @"kReaderPostServiceToggleSiteFollowingState";
```
Sure absolutely!

danielebogo commented on WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift:
```diff
@@ -144,18 +135,35 @@ extension ReaderTopicService {
+    ///   - failure: Failure block
+    func toggleSubscribingComments(for siteId: Int?, subscribe: Bool, _ success: (() -> Void)? = nil, _ failure: ((NSError?) -> Void)? = nil) {
+        guard let siteId = siteId else {
+            return
```
Yes you are right. I can call the block passing a custom error

danielebogo commented on WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift:
```diff
@@ -107,12 +81,29 @@ extension ReaderTopicService {
+    ///   - failure: Failure block
+    func toggleSubscribingNotifications(for siteId: Int?, subscribe: Bool, _ success: (() -> Void)? = nil, _ failure: ((NSError?) -> Void)? = nil) {
+        guard let siteId = siteId else {
+            return
```
Same here like my comment above

danielebogo commented on WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift:
```diff
@@ -181,16 +189,50 @@ extension ReaderTopicService {
+    ///   - failure: Failure block
+    func toggleSubscribingEmail(for siteId: Int?, subscribe: Bool, _ success: (() -> Void)? = nil, _ failure: ((NSError?) -> Void)? = nil) {
+        guard let siteId = siteId else {
+            return
```
Same here

danielebogo commented:
Hi @aerych you are perfectly right about the inconsistence. Now i changed the topic service extension in order to use the site id as Int rather than NSNumber as much as i can. But at the end a NSNumber is required due to the objective-c objects. Might be great if it can be aligned in all the places around the app, not only for the site bu also for the other models like Post for instance.

danielebogo commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationSettingsViewController.swift:
```diff
@@ -244,33 +284,87 @@ open class NotificationSettingsViewController: UIViewController {
+
+        switch section {
+        case .followedSites: return followedSites[index.row]
+        default: return nil
```
+1

danielebogo commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationSettingsViewController.swift:
```diff
@@ -244,33 +284,87 @@ open class NotificationSettingsViewController: UIViewController {
 
+        switch section {
+        case .followedSites: return followedSites.isEmpty
+        default: return groupedSettings?[section]?.count == 0
```
+1

aerych commented on WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift:
```diff
@@ -107,18 +81,37 @@ extension ReaderTopicService {
+    ///   - failure: Failure block
+    func toggleSubscribingNotifications(for siteId: Int?, subscribe: Bool, _ success: (() -> Void)? = nil, _ failure: ((NSError?) -> Void)? = nil) {
+        guard let siteId = siteId else {
+            let error = NSError(domain: "ReaderTopicService", code: 0, userInfo: [NSLocalizedDescriptionKey: "The site id is nil"])
```
Where we're creating Errors to pass to failure blocks, it might be nice if we could abstract the error creation so we could both encapsulate it, and to make it easier to add new errors in the future should the need arise.  You might take a peek at `SnapshotError` in `WordPress/WordPressScreenshotGeneration/SnapshotHelper.swift` for an example of one approach. 

Additionally, if we're using `NSLocalizedDescriptionKey` then the message should be localized as well.

danielebogo replied:
Perfect! I'll modify also The remote service using the same Error rather than NSError.

aerych commented:
Actually, before I forget...
Making a note that there is an issue with the version 75 of our core data model that we've been discussing in slack. Its unrelated to anything being worked on here but since this branch is using the version 75 we'll want to be careful how we merge back to develop.

danielebogo commented:
Hi @aerych i did some change. I created a `ReaderTopicServiceError` with the cases i need to handle into **WordPressKit**. I modified the remote service within WordPressKit to use the same Error. Also i replaced all the NSError and i improved the methods accepting an Int rather than NSNumber. I also fixed the Unit Tests within **WordPressKit**.
Let me know what you think ok?

aerych commented on WordPressKit/WordPressKit/ReaderTopicServiceError.swift:
```diff
@@ -0,0 +1,21 @@
+    public var description: String {
+        switch self {
+        case .invalidId:
+            return "Invalid id: an id must be valid or not nil"
```
We still need to localize the string that's returned by passing it to `NSLocalizedString`.  There are lots of examples throughout the app to look at. :)   Be sure to include a helpful comment about how the string is used so our translators will understand its context. :)

One thing to be aware of is that any parameters included in the string will not have their final values when the string is translated.  In this case, I'd suggest maybe concatenating the parameter's value to the end of the concatenated string. 

No need to worry about the actual translation, we have a separate process that handles all string translations.

aerych commented on WordPressKit/WordPressKit/ReaderTopicServiceRemote+Subscription.swift:
```diff
@@ -73,25 +73,25 @@ extension ReaderTopicServiceRemote {
+        }) { (error, response) in
             DDLogError("Error: \(error.localizedDescription)")
-            failure(error)
+            failure(ReaderTopicServiceError.remoteResponse(message: error.localizedDescription, url: response?.url?.absoluteString ?? "unknown response url"))
```
For easier readability could `response?.url?.absoluteString ?? "unknown response url"` be performed first and the result passed as an argument? 
Also, we might consider shortening the message to "unknown url" but either way we need to localize the string (NSLocalizedString).

aerych commented on WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift:
```diff
@@ -100,9 +98,9 @@ extension ReaderTopicService {
+        let failureBlock = { (error: ReaderTopicServiceError?) in
             failure?(error)
-            DDLogError("Error turn on notifications: \(error?.localizedDescription ?? "unknown error")")
+            DDLogError("Error turn on notifications: \(error?.description ?? "unknown error")")
```
👍

aerych commented:
I retested the feature after the last changes.  Everything works great.  I think once we get the strings taken care of we can :shipit: 
Nice work @danielebogo :)

danielebogo commented:
Hey @aerych i localized the errors strings. Let me know if the comments are ok :)

aerych commented:
Great job @danielebogo :D

danielebogo commented:
Thanks @aerych ! :)
